runtime/src/main/cpp/Memory.cpp

```diff
@@ -2504,6 +2567,99 @@ void shareAny(ObjHeader* obj) {
   container->makeShared();
 }
 
+OBJ_GETTER0(detectCyclicReferences) {
+  // Collect rootset, hold references to simplify remaining code.
+  KRefList rootset;
+  lock(&g_leakCheckerGlobalLock);
+  auto* candidate = g_leakCheckerGlobalList;
+  while (candidate != nullptr) {
+    CreateStablePointer(candidate);
+    rootset.push_back(candidate);
+    candidate = candidate->meta_object()->LeakDetector.next_;
+  }
+  unlock(&g_leakCheckerGlobalLock);
+  KRefSet cyclic;
+  KRefSet seen;
+  KRefDeque toVisit;
+  for (auto* root: rootset) {
+    seen.clear();
+    toVisit.clear();
+    traverseReferredObjects(root, [&toVisit](ObjHeader* obj) { toVisit.push_front(obj); });
+    bool seenToRoot = false;
+    while (!toVisit.empty() && !seenToRoot) {
+      KRef current = toVisit.front();
+      toVisit.pop_front();
+      if (current == root) seenToRoot = true;
+      // TODO: racy against concurrent mutators.
+      if (seen.count(current) == 0) {
+        traverseReferredObjects(current, [&toVisit](ObjHeader* obj) {
+           toVisit.push_front(obj);
+        });
+        seen.insert(current);
+      }
+    }
+    if (seenToRoot && cyclic.count(root) == 0) {
```

Comment:
Why check for `cyclic.count(root) == 0`?

Reply:
Debug code, removed.